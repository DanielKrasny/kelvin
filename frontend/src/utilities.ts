type Method = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';

/**
 * Get data from API
 * @param url url to fetch
 * @param data data which will be sent to the API
 * @param method HTTP method, if you want to override the default GET/POST
 * @note If data is passed, the request will be a POST request, otherwise a GET request, if not overridden by method parameter
 * @returns $ReturnType, if the request was successful, otherwise undefined
 */
export const getFromAPI = async <$ReturnType>(
    url: string,
    data?: unknown,
    method?: Method
): Promise<$ReturnType | undefined> => {
    try {
        const response = await fetch(url, {
            method: method || (data ? 'POST' : 'GET'),
            body: data ? JSON.stringify(data) : undefined
        });

        if (!response.ok) {
            return undefined;
        }

        const json = await response.json();
        return json as $ReturnType;
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (_) {
        return undefined;
    }
};

/**
 * Generate array with range of numbers from start
 * @param size size of the array
 * @param start starting number
 * @returns array of numbers from start to start + size
 */
export const generateRange = (size: number, start = 0) => {
    return Array.from({ length: size }).map((_, i) => i + start);
};

/**
 * Generate string date format from date
 * @param date Date object or string in format 'YYYY-MM-DD' or 'YYYY-MM-DDTHH:MMZ'
 * @param withTime Include time in the output
 * @returns If withTime is false 'YYYY-MM-DD', otherwise 'YYYY-MM-DD HH:MM'
 */
export const getDate = (date: Date | string, withTime = false) => {
    let d: Date;
    if (typeof date === 'string') {
        d = new Date(date);
    } else {
        d = date;
    }

    const year = d.getFullYear();
    const month = (d.getMonth() + 1).toString().padStart(2, '0');
    const day = d.getDate().toString().padStart(2, '0');

    if (!withTime) {
        return `${year}-${month}-${day}`;
    }

    const hours = d.getHours().toString().padStart(2, '0');
    const minutes = d.getMinutes().toString().padStart(2, '0');
    return `${year}-${month}-${day} ${hours}:${minutes}`;
};
