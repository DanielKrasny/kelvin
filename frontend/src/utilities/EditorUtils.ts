import CodeMirror from 'codemirror';

// Remove string Type Union from EditorConfiguration
export type ExtraKeys<$BaseType> = $BaseType extends string ? never : $BaseType;

export type Arrayable<$Type> = $Type | $Type[];

export const inArrayable = <T>(arrayable: Arrayable<T>, value: T) => {
    if (Array.isArray(arrayable)) {
        return arrayable.some((item) => item === value);
    }
    return arrayable === value;
};

export const arrayify = <T>(arrayable: Arrayable<T>) => {
    if (Array.isArray(arrayable)) return arrayable;
    return [arrayable];
};

/**
 * Type definition for extension
 * Each extension is a function. This function can accept these parameters:
 * Extension function is the main code for extension. You can accept these parameters:
 * - fileName - current file filename
 * - code - current code in editor
 * - editor - current editor instance
 * - type - gives you information about context, in which was function called. It can be one of these:
 *   - setup - This is called when editor is created (first time), or filename changed
 *   - lint - This is called when linting is requested (after code change)
 *   - hint - This is called when hinting is requested (after user presses ctrl+space)
 *
 * From this function you can return these things:
 * - Undefined for case, where you don't care about for example current file/extension (extension
 *   for json file doesn't do anything with python file, etc..), empty object is also valid return value
 *   and do same thing as undefined
 * - Object containing these keys:
 *   - spellCheck - Controlls if spellcheck should be enabled/disabled (only for setup)
 *   - gutters - List of gutters to set for current editor (only for setup)
 *   - hint - This will be array of hints, for user to select them
 *   - lint - Lint warning/errors for current code
 * - You don't need to set every key in object, only these, you do care. For example:
 *   - if you only care for hints, return something like this:
 *     ```TS
 *     return {
 *        hint: {
 *            list: ["const", "console"],
 *            from: {
 *               ch: 2,
 *               line: 2
 *            },
 *            to: {
 *                ch: 5,
 *                line: 2
 *            }
 *        }
 *     }
 *     ```
 *   - or if you want to enable spellcheck on setup:
 *     ```TS
 *     return {
 *        spellCheck: true
 *     }
 *     ```
 */
export type EditorExtension = (
    fileName: string,
    code: string,
    editor: CodeMirror.Editor,
    type: 'setup' | 'lint' | 'hint'
) =>
    | undefined
    | {
          spellCheck?: boolean;
          gutters?: string[];
          hint?: CodeMirror.Hint;
          lint?: CodeMirror.LintError[];
      };

export const getExtension = (filename: string) => {
    if (!filename) return;

    const parts = filename.split('.');
    return parts[parts.length - 1].toLowerCase();
};
