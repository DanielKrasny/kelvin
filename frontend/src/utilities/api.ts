export function csrfToken() {
    return document.querySelector('meta[name=csrf-token]').getAttribute('content');
}
