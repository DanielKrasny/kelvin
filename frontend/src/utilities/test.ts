import { vi } from 'vitest';

/**
 * Mock fetch requests
 * Currently accept one parameter, which is object, mapping url to
 * response object, which consists of okStatus and return data
 * returned by json() method
 */
export const mockFetch = (
    urlsMap: Record<
        string,
        {
            okStatus: boolean;
            data: unknown;
        }
    >
) => {
    const keys = Object.keys(urlsMap);

    vi.stubGlobal(
        'fetch',
        vi.fn((url: string) => {
            if (!keys.includes(url)) {
                return Promise.resolve({
                    ok: false,
                    json: () => Promise.resolve(undefined)
                });
            }

            const resultData = urlsMap[url];

            return Promise.resolve({
                ok: resultData.okStatus,
                json: () => Promise.resolve(resultData.data)
            });
        })
    );
};
