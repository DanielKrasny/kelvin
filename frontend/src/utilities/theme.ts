import { effect, ref } from 'vue';
import { localStorageStore } from './storage';

export type ThemeValue = 'auto' | 'light' | 'dark';
export const currentTheme = localStorageStore<ThemeValue>('theme', 'auto');

//this ref should be used for the current set theme in other parts of application
export const theme = ref<'light' | 'dark'>('light');

//function which detects current theme from media query
const htmlData = document.querySelector('html').dataset;
const themeMedia = window.matchMedia('(prefers-color-scheme: light)');

function detectTheme() {
    theme.value = themeMedia.matches ? 'light' : 'dark';
}

//since we want to have listener on media query to update theme
//every time when user changes theme in browser, we can't use
//computed, and we need to use effect
effect(() => {
    if (currentTheme.value === 'auto') {
        themeMedia.addEventListener('change', detectTheme);
        detectTheme();
        return;
    }
    //in future if this will became SPA, and ColorTheme component will not
    //be presented on every page we should call this in onUnmounted
    themeMedia.removeEventListener('change', detectTheme);

    theme.value = currentTheme.value;
});

//every time theme changes we want to update dataset in
//html tag for bootstrap to change theme
effect(() => {
    htmlData.bsTheme = theme.value;
});
