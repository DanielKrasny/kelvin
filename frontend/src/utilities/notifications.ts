import { getDataWithCSRF } from './api';

//@TODO: complete null unions if missing
export type Notification = {
    id: number;
    level: 'info'; //@TODO: add other levels
    recepient: number;
    unread: boolean;
    actor_content_type: number;
    actor_object_id: string;
    verb: string;
    description: string | null;
    timestamp: string;
    public: boolean;
    deleted: boolean;
    emailed: boolean;
    important: boolean;
    actor: string;
    action_object: string;
    action_object_url: string;
    custom_text?: string;
    target?: string;
};

export const getNotifications = async () => {
    const data = await getDataWithCSRF<{ notifications: Notification[] }>('/notification/all');
    return data.notifications;
};

export const markRead = async (id: number) => {
    const data = await getDataWithCSRF<{ notifications: Notification[] }>(
        '/notification/mark_as_read/' + id,
        'POST'
    );
    return data.notifications;
};

export const markAllRead = async () => {
    const data = await getDataWithCSRF<{ notifications: Notification[] }>(
        '/notification/mark_as_read',
        'POST'
    );
    return data.notifications;
};

const getPublicKey = () => {
    return document.querySelector<HTMLMetaElement>('meta[name="django-webpush-vapid-key"]').content;
};

const urlB64ToUint8Array = (base64String: string) => {
    const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
    const base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');

    const rawData = window.atob(base64);
    const outputArray = new Uint8Array(rawData.length);

    for (let i = 0; i < rawData.length; ++i) {
        outputArray[i] = rawData.charCodeAt(i);
    }
    return outputArray;
};

const getSubscription = async (reg: ServiceWorkerRegistration) => {
    const subscribeOpts = {
        userVisibleOnly: true,
        applicationServerKey: urlB64ToUint8Array(getPublicKey())
    };

    try {
        let sub = await reg.pushManager.getSubscription();
        if (sub) {
            return sub;
        }

        sub = await reg.pushManager.subscribe(subscribeOpts);

        const browser = navigator.userAgent
            .match(/(firefox|msie|chrome|safari|trident)/gi)[0]
            .toLowerCase();
        const data = {
            status_type: 'subscribe',
            subscription: sub.toJSON(),
            browser: browser,
            group: null
        };

        await getDataWithCSRF('/webpush/save_information', 'POST', data, {
            'Content-Type': 'application/json'
        });
        return sub;
    } catch (exception) {
        console.log(exception);
    }

    return null;
};

export const subscribePushNotifications = async (reg: ServiceWorkerRegistration) => {
    const isEnabled = (await getSubscription(reg)) !== null;
    return isEnabled;
};

export type PushNotifications =
    | {
          supported: true;
          enabled: boolean;
          reg: ServiceWorkerRegistration;
      }
    | {
          supported: false;
          enabled: null;
          reg: null;
      };

export const getPushStatus = async (): Promise<PushNotifications> => {
    let reg: null | ServiceWorkerRegistration = null;
    if ('serviceWorker' in navigator && 'PushManager' in window && getPublicKey()) {
        try {
            reg = await navigator.serviceWorker.register('/static/service-worker.js');
            if (!reg.showNotification) {
                return;
            }
            return {
                supported: true,
                enabled: await subscribePushNotifications(reg),
                reg
            };
        } catch (err) {
            console.log(err);
            return {
                supported: true,
                enabled: false,
                reg
            };
        }
    }

    return {
        supported: false,
        enabled: null,
        reg: null
    };
};
