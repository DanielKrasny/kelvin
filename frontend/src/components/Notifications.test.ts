import { mount } from '@vue/test-utils';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import Notifications from './Notifications.vue';
import { mockFetch } from '../utilities/test';
import { Notification } from '../utilities/notifications';

const exampleNotifications = [
    {
        id: 1,
        action_object: 'some-action-object',
        action_object_url: 'some-action-object-url',
        actor: 'some-actor1',
        actor_content_type: 1,
        actor_object_id: 'some-actor-object-id',
        custom_text: 'some-custom-text',
        deleted: false,
        description: 'some-description',
        emailed: false,
        important: false,
        level: 'info',
        public: false,
        recepient: 1,
        timestamp: '0',
        unread: false,
        verb: 'some-verb'
    },
    {
        id: 2,
        action_object: 'some-action-object',
        action_object_url: 'some-action-object-url2',
        actor: 'some-actor2',
        actor_content_type: 1,
        actor_object_id: 'some-actor-object-id2',
        custom_text: 'some-custom-text2',
        deleted: false,
        description: 'some-description2',
        emailed: false,
        important: false,
        level: 'info',
        public: false,
        recepient: 1,
        timestamp: '0',
        unread: false,
        verb: 'some-verb'
    }
] satisfies Notification[];

beforeEach(() => {
    //fake csrf token
    const csrf = document.createElement('meta');
    csrf.name = 'csrf-token';
    csrf.content = 'some-token';
    document.head.appendChild(csrf);

    //fake vapid key
    const vapid = document.createElement('meta');
    vapid.name = 'django-webpush-vapid-key';
    vapid.content = 'some-key';
    document.head.appendChild(vapid);

    mockFetch({
        '/notification/all': {
            okStatus: true,
            data: {
                notifications: exampleNotifications
            }
        },
        '/notification/mark_as_read/1': {
            okStatus: true,
            data: {
                notifications: exampleNotifications.slice(1)
            }
        }
    });
});

afterEach(() => {
    document.querySelectorAll("meta[name='csrf-token']").forEach((e) => e.remove());
    document.querySelectorAll("meta[name='django-webpush-vapid-key']").forEach((e) => e.remove());

    vi.restoreAllMocks();
});

describe('Notifications', () => {
    it("Don't wait for notifications", () => {
        const wrapper = mount(Notifications);

        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Wait for notification', async () => {
        const wrapper = mount(Notifications);

        await vi.waitFor(() => expect(wrapper.html()).toContain('some-actor1'));
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Read notification', async () => {
        const fetchSpy = vi.spyOn(globalThis, 'fetch');

        const wrapper = mount(Notifications);

        await vi.waitFor(() => expect(wrapper.html()).toContain('some-actor1'));

        await wrapper.find('.btn-close').trigger('click');

        await vi.waitFor(() => expect(wrapper.html()).not.toContain('some-actor1'));

        expect(fetchSpy).toBeCalledWith(
            '/notification/mark_as_read/' + exampleNotifications[0].id,
            expect.anything()
        );

        expect(wrapper.html()).toMatchSnapshot();
    });
});
