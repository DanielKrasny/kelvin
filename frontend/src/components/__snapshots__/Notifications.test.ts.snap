// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notifications > Don't wait for notifications 1`] = `
"<li class="nav-item dropdown"><button class="btn nav-link dropdown-toggle" href="#" type="button" data-bs-toggle="dropdown" data-bs-auto-close="outside" aria-expanded="false" title="Notifications"><span class="iconify" data-icon="bi:bell"></span><span class="d-md-none ms-1">Notifications</span>
    <!--v-if-->
  </button>
  <div class="dropdown-menu dropdown-menu-end shadow p-0 rounded dropdown-menu-custom">
    <ul class="list-group list-group-flush" style="max-height: 50vh; overflow-y: auto;">
      <li class="list-group-item">
        <div class="d-flex align-items-center">
          <div> Notifications
            <!--v-if-->
          </div>
          <div class="ms-auto">
            <!--v-if--><button class="btn text-body text-muted" title="Clear all notifications"><span class="iconify" data-icon="mdi:notification-clear-all"></span></button>
          </div>
        </div>
      </li><span class="list-group-item p-1 text-center">There are no notifications!</span>
    </ul>
  </div>
</li>"
`;

exports[`Notifications > Read notification 1`] = `
"<li class="nav-item dropdown"><button class="btn nav-link dropdown-toggle" href="#" type="button" data-bs-toggle="dropdown" data-bs-auto-close="outside" aria-expanded="false" title="Notifications"><span class="iconify" data-icon="bi:bell"></span><span class="d-md-none ms-1">Notifications</span>
    <!--v-if-->
  </button>
  <div class="dropdown-menu dropdown-menu-end shadow p-0 rounded dropdown-menu-custom">
    <ul class="list-group list-group-flush" style="max-height: 50vh; overflow-y: auto;">
      <li class="list-group-item">
        <div class="d-flex align-items-center">
          <div> Notifications
            <!--v-if-->
          </div>
          <div class="ms-auto">
            <!--v-if--><button class="btn text-body text-muted" title="Clear all notifications"><span class="iconify" data-icon="mdi:notification-clear-all"></span></button>
          </div>
        </div>
      </li>
      <li class="list-group-item p-1 d-flex align-items-center justify-content-between text-body-secondary">
        <div><strong>some-actor2&nbsp;</strong>
          <div>some-custom-text2</div><span>&nbsp;(<time datetime="1999-12-31T23:00:00.000Z" title="1. 1. 2000 0:00:00"><!--v-if-->25 years ago<!--v-if--></time>) </span>
        </div>
        <div><button type="button" hidden="" class="btn-close" aria-label="Close"></button></div>
      </li>
    </ul>
  </div>
</li>"
`;

exports[`Notifications > Wait for notification 1`] = `
"<li class="nav-item dropdown"><button class="btn nav-link dropdown-toggle" href="#" type="button" data-bs-toggle="dropdown" data-bs-auto-close="outside" aria-expanded="false" title="Notifications"><span class="iconify" data-icon="bi:bell"></span><span class="d-md-none ms-1">Notifications</span>
    <!--v-if-->
  </button>
  <div class="dropdown-menu dropdown-menu-end shadow p-0 rounded dropdown-menu-custom">
    <ul class="list-group list-group-flush" style="max-height: 50vh; overflow-y: auto;">
      <li class="list-group-item">
        <div class="d-flex align-items-center">
          <div> Notifications
            <!--v-if-->
          </div>
          <div class="ms-auto">
            <!--v-if--><button class="btn text-body text-muted" title="Clear all notifications"><span class="iconify" data-icon="mdi:notification-clear-all"></span></button>
          </div>
        </div>
      </li>
      <li class="list-group-item p-1 d-flex align-items-center justify-content-between text-body-secondary">
        <div><strong>some-actor1&nbsp;</strong>
          <div>some-custom-text</div><span>&nbsp;(<time datetime="1999-12-31T23:00:00.000Z" title="1. 1. 2000 0:00:00"><!--v-if-->25 years ago<!--v-if--></time>) </span>
        </div>
        <div><button type="button" hidden="" class="btn-close" aria-label="Close"></button></div>
      </li>
      <li class="list-group-item p-1 d-flex align-items-center justify-content-between text-body-secondary">
        <div><strong>some-actor2&nbsp;</strong>
          <div>some-custom-text2</div><span>&nbsp;(<time datetime="1999-12-31T23:00:00.000Z" title="1. 1. 2000 0:00:00"><!--v-if-->25 years ago<!--v-if--></time>) </span>
        </div>
        <div><button type="button" hidden="" class="btn-close" aria-label="Close"></button></div>
      </li>
    </ul>
  </div>
</li>"
`;
